### PR TITLE
Support multiple headers sharing the same name

### LINKED ISSUE #130: can't use multiple HTTP Headers with the same name
In HTTP it allows multiple headers with the same name.  However, when I went to test
webapp that seems to break when IE10 sends multiple cookie headers, I can't use httpie to test it as it only puts in the last header value in the request.

example:

```bash
$ http -v example.org Cookie:foo1=bar1 User-Agent:BAM\1.0 Cookie:foo2=bar2
```

Current output:

```http
GET / HTTP/1.1
Accept: _/_
Accept-Encoding: gzip, deflate, compress
Cookie: foo2=bar2
Host: example.org
User-Agent: BAM1.0

//response omitted
```

Expected output:

```
GET / HTTP/1.1
Accept: _/_
Accept-Encoding: gzip, deflate, compress
Cookie: foo1=bar1
Cookie: foo2=bar2
Host: example.org
User-Agent: BAM1.0
```

Really HTTP headers should be represented as a case-insensitive MultiMap of some sort, not a  Dict.

### PR DESCRIPTION
**Pull request by isidentical**
May resolve #130. If the design is agreed upon, may require change in a few places:
- Changelog
- Build scripts (not sure about this, but I guess adding a dependency would mean I need to change a bunch of places in `docs/packaging`).

### CODE REVIEW

File: httpie/client.py [hunk 1]
```diff
@@ -190,10 +191,37 @@ def finalize_headers(headers: RequestHeadersDict) -> RequestHeadersDict:
...
+            # If the current value is not among the initial values
+            # set for this field, then it means that this field got
+            # overriden on the way, and we should preserve it.
+            continue
+
+        new_headers.popone(name)
+        for value in original_values:
+            new_headers.add(name, value)
+
+    prepared_request.headers = new_headers
```
Comment (jakubroztocil) [inline httpie/client.py:222]: Here we use the name `value` for both the original and prepared values inside the same loop. Let’s prefix the names to reflect the difference and make it easier to follow.

**Review by jakubroztocil — CHANGES_REQUESTED** [review]
Looks great 👍🏻  Just ~one~ a couple of inline comments & also noticed that with mixed-case names, all the resulting names are the same, but we should preserve the original names verbatim:

```bash
$ http --offline pie.dev/get AAA:AAA aaa:aaa
```

```http
GET /get HTTP/1.1
aaa: AAA   ← name should stay "AAA"
aaa: aaa
```

File: httpie/client.py [hunk 2]
```diff
@@ -190,10 +191,37 @@ def finalize_headers(headers: RequestHeadersDict) -> RequestHeadersDict:
...
 
+def apply_missing_repeated_headers(
+    prepared_request: requests.PreparedRequest,
+    original_headers: RequestHeadersDict
+) -> None:
+    """Update the given `prepared_request`'s headers with the original
+    ones. This allows the requests to be prepared as usual, and then later
+    merged with headers that are specified multiple times."""
+
+    new_headers = RequestHeadersDict(prepared_request.headers)
```
Comment (jakubroztocil) [inline httpie/client.py:206]: Couldn’t we just `.copy()` the headers?

File: docs/README.md [hunk 3]
```diff
@@ -869,6 +869,41 @@ To send a header with an empty value, use `Header;`, with a semicolon:
...
+If the request is sent with multiple headers that are sharing the same name, then
+the HTTPie will send them individually.
+
+```bash
+http --offline example.org Cookie:one Cookie:two
+```
+
+```http
+GET / HTTP/1.1
+
```
Comment (jakubroztocil) [inline docs/README.md:886]:
```suggestion
```

File: docs/README.md [hunk 4]
```diff
@@ -869,6 +869,41 @@ To send a header with an empty value, use `Header;`, with a semicolon:
...
+It is also possible to pass a single header value pair, where the value is a comma
+separated list of header values. Then the client will send it as a single header.
+
+```bash
+http --offline example.org Numbers:one,two
+```
+
+```
+GET / HTTP/1.1
+
```
Comment (jakubroztocil) [inline docs/README.md:900]:
```suggestion
```

File: docs/README.md [hunk 5]
```diff
@@ -869,6 +869,41 @@ To send a header with an empty value, use `Header;`, with a semicolon:
...
+```
+
+It is also possible to pass a single header value pair, where the value is a comma
+separated list of header values. Then the client will send it as a single header.
+
+```bash
+http --offline example.org Numbers:one,two
+```
+
+```
```
Comment (jakubroztocil) [inline docs/README.md:898]: Missing `http` as the languge type

File: httpie/client.py [hunk 6]
```diff
@@ -190,10 +191,37 @@ def finalize_headers(headers: RequestHeadersDict) -> RequestHeadersDict:
...
+    for name, value in prepared_request.headers.items():
+        if name not in original_headers:
+            continue
+
+        original_values = original_headers.getall(name)
+        if value not in original_values:
+            # If the current value is not among the initial values
+            # set for this field, then it means that this field got
+            # overriden on the way, and we should preserve it.
```
Comment (jakubroztocil) [inline httpie/client.py:215]:
```suggestion
            # overridden on the way, and we should preserve it.
```

File: httpie/client.py (hunk 2 above)
Comment (isidentical) [reply httpie/client.py:206]: `prepared_request.headers` is a regular dict, we are casting it into a `RequestHeadersDict`.

File: docs/README.md (hunk 5 above)
Comment (isidentical) [reply docs/README.md:898]:
```suggestion
```http
```

**Comment by jakubroztocil** [comment]
We should also add a CHANGELOG entry.

**Comment by jakubroztocil** [comment]
The `packit` error is because of the new dependency (not sure how to prevent that). 

But the other failing checks should be fixed. 

(And it looks like I [have to run them manually](https://github.blog/changelog/2021-04-22-github-actions-maintainers-must-approve-first-time-contributor-workflow-runs/) for this first PR.)

**Comment by isidentical** [comment]
> (And it looks like I have to run them manually for this first PR.)

I think it should always run for me if you change this (`Settings > Actions`):
image

**Comment by jakubroztocil** [comment]
✅ settings changed

**Comment by isidentical** [comment]
> The packit error is because of the new dependency (not sure how to prevent that).

I am also not sure about it. Seems like `fedora` package for `multidict` is `4.7.6` (https://fedora.pkgs.org/33/fedora-x86_64/python3-multidict-4.7.6-3.fc33.x86_64.rpm.html) but I've specified `5.0.0` on the `setup.py` (not that we explicitly needed it, just seemed like a stable release). I've downgraded it to `4.7.0` and will test that way.

**Comment by isidentical** [comment]
Yep, seems to be working now. The only thing that doesn't work is the markdown check which I don't think is related to this PR. It is showing line 34 from the changelog (this PR only touches line 8-10), and line 1950 from docs (this PR touches 850-900). Any idea about what is going on with that action? @jakubroztocil

File: httpie/client.py (hunk 1 above)
Comment (isidentical) [reply httpie/client.py]: `prepared_value`/`original_value` distinction is now added.

**Review by jakubroztocil — CHANGES_REQUESTED** [review]
Looks great. Last thing: can we add tests mixed-case names scenario?

**Comment by jakubroztocil** [comment]
Oh and the Markdown liner is yelling something about blank lines: https://github.com/httpie/httpie/runs/4055683412?check_suite_focus=true#step:5:9

**Comment by isidentical** [comment]
> Looks great. Last thing: can we add tests mixed-case names scenario?

We already do (the offline test implicitly includes that).

> I think the Markdown linter just complains about too many blank lines. The action might be misconfigured that it didn’t get run when the error on line 34 got committed. But please fix it.

I am sorry but I can't seem to find out where are double blank lines. 

```
    29  - Fixed `--verbose` HTTP 307 redirects with streamed request body. ([#1088](https://github.com/httpie/httpie/issues/1088))
    30  - Fixed handling of session files with `Cookie:` followed by other headers. ([#1126](https://github.com/httpie/httpie/issues/1126))
    31
    32  ## [2.4.0](https://github.com/httpie/httpie/compare/2.3.0...2.4.0) (2021-02-06)
    33
    34  - Added support for `--session` cookie expiration based on `Set-Cookie: max-age=<n>`. ([#1029](https://github.com/httpie/httpie/issues/1029))
    35  - Show a `--check-status` warning with `--quiet` as well, not only when the output is redirected. ([#1026](https://github.com/httpie/httpie/issues/1026))
    36  - Fixed upload with `--session` ([#1020](https://github.com/httpie/httpie/issues/1020)).
    37  - Fixed a missing blank line between request and response ([#1006](https://github.com/httpie/httpie/issues/1006)).
    38
    39  ## [2.3.0](https://github.com/httpie/httpie/compare/2.2.0...2.3.0) (2020-10-25)
```

As you can see there are no double-blank lines.

**Comment by jakubroztocil** [comment]
> > Looks great. Last thing: can we add tests mixed-case names scenario?
> 
> We already do (the offline test implicitly includes that).
> 

Got it!

> > I think the Markdown linter just complains about too many blank lines. The action might be misconfigured that it didn’t get run when the error on line 34 got committed. But please fix it.
> 
> I am sorry but I can't seem to find out where are double blank lines.
> 
> ```
>     29  - Fixed `--verbose` HTTP 307 redirects with streamed request body. ([#1088](https://github.com/httpie/httpie/issues/1088))
>     30  - Fixed handling of session files with `Cookie:` followed by other headers. ([#1126](https://github.com/httpie/httpie/issues/1126))
>     31
>     32  ## [2.4.0](https://github.com/httpie/httpie/compare/2.3.0...2.4.0) (2021-02-06)
>     33
>     34  - Added support for `--session` cookie expiration based on `Set-Cookie: max-age=<n>`. ([#1029](https://github.com/httpie/httpie/issues/1029))
>     35  - Show a `--check-status` warning with `--quiet` as well, not only when the output is redirected. ([#1026](https://github.com/httpie/httpie/issues/1026))
>     36  - Fixed upload with `--session` ([#1020](https://github.com/httpie/httpie/issues/1020)).
>     37  - Fixed a missing blank line between request and response ([#1006](https://github.com/httpie/httpie/issues/1006)).
>     38
>     39  ## [2.3.0](https://github.com/httpie/httpie/compare/2.2.0...2.3.0) (2020-10-25)
> ```
> 
> As you can see there are no double-blank lines.

Strange, I also get no error when running the check locally (`sudo gem install mdl; make doc-check`). Let’s merge and fix later.

**Comment by isidentical** [comment]
> Strange, I also get no error when running the check locally (sudo gem install mdl; make doc-check). Let’s merge and fix later.

Ah, got it. I wasn't able to see it locally since I did not rebase. And this problem got introduced in your 2.6.0 release commit.

**Comment by isidentical** [comment]
Once I rebased, I was able to see it but yes it can be fixed later. Thanks for merging!

**Comment by jakubroztocil** [comment]
Thanks for the PR — it’s a great improvement 🤘🎉 

(Let’s consider bumping the major version. We have another mildly breaking change in the making https://github.com/httpie/httpie/pull/1169, so we can bundle them together in the next release.)

**Pull request merged**